File: roboticaSourceCode/calculator.cpp
#include <iostream>
#include "roboticaFunctions.h"
#define _USE_MATH_DEFINES

#include <math.h>
#include <cmath>


/*
				ROBOT ASCII FOR VARIABLE EXPLANATION

			   base		   elbow	  camera    |  
				O--------------O-----------[]---O gripperbase
				|   humerus		 ulna		   |
				|							  /\  gripper
				|
				|
			  __|__

			O = joint
			-,| = solid
			[] = special component
*/

//constants
const double DEGREES_TO_RADIANS = M_PI / 180.0; //multiplication factor to turn degree values into radian values

//variables 
double base_rotation_angle = 90; // Angle between base and humerus in degrees
double elbow_rotation_angle = 0; // Angle between humerus and ulna in degrees
double length_humerus = 40.5; // Length of the humerus in units (cm)
double length_ulna = 32.0; // Length of the ulna in units (cm)
double camera_position_on_ulna = 30.0; // Distance from elbow joint to camera on ulna in units
double camera_height_from_ground = 49.0; // Height of the camera from the ground in units

//camera specifications
const double camera_fov = 61.0; // Diagonal FOV in degrees
const int camera_width = 640; // Camera resolution width in pixels
const int camera_height = 480; // Camera resolution height in pixels


//calculate the real-life coordinates of the object
Point calculateRealLifeCoordinates(int screen_x, int screen_y) {
	// Convert screen coordinates to normalized coordinates (-1 to 1)
	// The normalised x uses the y coordiantes and vice versa. if you dont do this the read coordinates are flipped 90 degrees for some reason
	double normalized_y = (2.0 * screen_x / camera_width) - 1.0;
	double normalized_x = 1.0 - (2.0 * screen_y / camera_height);

	//calculate the camera's horizontal and vertical FOV in radians
	double diagonal_fov_radians = camera_fov * DEGREES_TO_RADIANS;
	double aspect_ratio = static_cast<double>(camera_width) / camera_height;
	double horizontal_fov_radians = 2 * atan(tan(diagonal_fov_radians / 2) / sqrt(1 + aspect_ratio * aspect_ratio));
	double vertical_fov_radians = 2 * atan(tan(diagonal_fov_radians / 2) / sqrt(1 + (1 / aspect_ratio) * (1 / aspect_ratio)));

	//calculate the real-life coordinates in the camera's frame of reference
	double camera_x = 2 * normalized_x * tan(horizontal_fov_radians / 2) * camera_position_on_ulna;
	double camera_y = 2 * normalized_y * tan(vertical_fov_radians / 2) * camera_position_on_ulna;

	//transform the coordinates to the robot's base frame of reference
	double humerus_angle_radians = base_rotation_angle * DEGREES_TO_RADIANS;
	double elbow_angle_radians = elbow_rotation_angle * DEGREES_TO_RADIANS;

	//position of the elbow joint
	Point elbow;
	elbow.x = length_humerus * cos(humerus_angle_radians);
	elbow.y = length_humerus * sin(humerus_angle_radians);

	//position of the wrist joint (end of ulna)
	Point wrist;
	wrist.x = elbow.x + length_ulna * cos(humerus_angle_radians + elbow_angle_radians);
	wrist.y = elbow.y + length_ulna * sin(humerus_angle_radians + elbow_angle_radians);

	//adjust the coordinates for the camera's position on the ulna
	Point camera_mount;
	camera_mount.x = elbow.x + camera_position_on_ulna * cos(humerus_angle_radians + elbow_angle_radians);
	camera_mount.y = elbow.y + camera_position_on_ulna * sin(humerus_angle_radians + elbow_angle_radians);

	//calculate the object position relative to the camera
	double object_x_relative = camera_x * cos(humerus_angle_radians + elbow_angle_radians) - camera_y * sin(humerus_angle_radians + elbow_angle_radians);
	double object_y_relative = camera_x * sin(humerus_angle_radians + elbow_angle_radians) + camera_y * cos(humerus_angle_radians + elbow_angle_radians);

	//absolute position of the object
	Point object;
	object.x = camera_mount.x + object_x_relative;
	object.y = camera_mount.y + object_y_relative;

	std::cout << object.x << " " << object.y << std::endl; //debug log
	return object;
}

//calculate the distance between the base and the target coordinates
double calculateDistanceToTarget(int x,int y){
	double distanceFromBaseToTarget = sqrt(pow(x, 2) + pow(y, 2)); //pythagoras' theorem
	return distanceFromBaseToTarget;
}

//calculate the needed rotation of the gripper to grab the object while remaining perpendicular to it
#include <cmath>

double calculateGripperAngle(double vx, double vy) {
    // Perpendicular direction vector
    double px = -vy;
    double py = vx;

    // Calculate the angle with respect to the x-axis
    double angleRadians = std::atan2(py, px);
    double angleDegrees = angleRadians * (180.0 / M_PI);

    // Normalize angle to be within the range of [-180, 180]
    if (angleDegrees > 180) {
        angleDegrees -= 360;
    } else if (angleDegrees < -180) {
        angleDegrees += 360;
    }

    // Account for the dead zone of the servos (<-150 and >150)
    if (angleDegrees > 150) {
        angleDegrees -= 180; // Flip 180 degrees
    } else if (angleDegrees < -150) {
        angleDegrees += 180; // Flip 180 degrees
    }

    // Normalize angle again to be within the range of [-150, 150]
    if (angleDegrees > 150) {
        angleDegrees -= 360;
    } else if (angleDegrees < -150) {
        angleDegrees += 360;
    }

    return angleDegrees;
}


//calculate the angles of the robot's elbow and base
double* calculateArmAngles(double x,double y,double distanceToTarget){ //x and y of the center of mass of the object, along with the distance.
	double* armAngles = new double[2]; //create a pointer towards an array with two decimal numbers
	//use trigonometry to calculate the angles: see https://www.geogebra.org/calculator/bf4wfqr5 for elaboration
	armAngles[0] = fmod((acos((pow(length_humerus, 2) + pow(length_ulna, 2) - pow(distanceToTarget, 2)) / (2 * length_humerus * length_ulna)) + M_PI), (2 * M_PI))*(180/M_PI);//elbow
	armAngles[1] = fmod((atan2(y, x) - atan2((length_ulna * sin(armAngles[0])), (length_humerus + length_ulna * cos(armAngles[0])))), (2 * M_PI))*(180/M_PI);//base angle, mult by gear ratio
	return armAngles; 
}

//run all the previous calculation
double* calculator(double x, double y, double vx, double vy) {
	double* results = new double[3];
	std::cout << "external c++ calculator has run with parameters " << x << " " << y << " " << vx << " " << vy << " \n"; //debug log

	Point pointC = calculateRealLifeCoordinates(x,y); //the point where gripper is going to be heading
	double distanceToTarget = calculateDistanceToTarget(pointC.x,pointC.y); //calculate distance from base to target coordinates
	// if(distanceToTarget > length_humerus + length_ulna || distanceToTarget < length_humerus - length_ulna){
	// 	std::cout << "\nout of reach\n" << std::endl; //cannot reach the point, return previous, assume object didnt move.
	// 	return results; //results of previous iteration
	// }
	double* armAngles = calculateArmAngles(pointC.x,pointC.y,distanceToTarget); // calculate required angles for the elbow and base servos
	results[0] = calculateGripperAngle(vx,vy);	//calculate angle for the gripper to be perpendicular to the scissors
	results[1] = armAngles[0]; //store the arm angle variables
	results[2] = armAngles[1];
	return results;
}

File: roboticaSourceCode/python/convertedFromCpp/calculator.py
import math
import numpy as np
from getServoPositions import getPosition
# Constants
DEGREES_TO_RADIANS = math.pi / 180.0  # Multiplication factor to turn degree values into radian values


"""
				ROBOT ASCII FOR VARIABLE EXPLANATION

			   base		   elbow	  camera	|  
				O--------------O-----------[]---O gripperbase
				|   humerus		 ulna			|
				|							   /\  gripper
				|
				|
			  __|__

			O = joint
			-,| = solid
			[] = special component
"""
# Variables


length_humerus = 40.5  # Length of the humerus in units (cm)
length_ulna = 32.0  # Length of the ulna in units (cm)
camera_position_on_ulna = 20.0  # Distance from elbow joint to camera on ulna in units
camera_height_from_ground = 43.0  # Height of the camera from the ground in units

# Camera specifications
camera_fov = 61.0  # Diagonal FOV in degrees
camera_width = 640  # Camera resolution width in pixels
camera_height = 480  # Camera resolution height in pixels

id_base = 2
id_elbow = 18
class Point:
	def __init__(self, x=0.0, y=0.0):
		self.x = x
		self.y = y


def calculate_camera_coords():
	base_rotation_angle = getPosition(id_base) / 3.33 * (math.pi/180) # Angle between base and humerus in degrees
	elbow_rotation_angle = getPosition(id_elbow) / 2 * (math.pi/180) # Angle between humerus and ulna in degrees
	
	elbow_coords = Point(length_humerus*math.cos(base_rotation_angle),length_humerus*math.sin(base_rotation_angle))

	elbow_theta = base_rotation_angle + elbow_rotation_angle
	camera_coords = Point(elbow_coords.x + (camera_position_on_ulna*math.cos(elbow_theta)),elbow_coords.y + (camera_position_on_ulna*math.sin(elbow_theta)))
	return [camera_coords,elbow_theta]


# Calculate the real-life coordinates of the object
def calculate_real_life_coordinates(screen_x, screen_y):
	data = calculate_camera_coords()
	camera_coords = data[0]
	elbow_theta = data[1]
	#print(f"camera: {camera_coords.x} {camera_coords.y}")

	conversion_rate = 38/640 #pixels per cm
	
	distance_camera_centroid = math.sqrt((-screen_x+320)**2+(screen_y-240)**2) * conversion_rate
	#print(f"distance: {distance_camera_centroid}")
	#print(distance_camera_centroid)
	theta_centroid = math.atan2((screen_x-320),(-screen_y+240))

	theta_centroid_real_life = theta_centroid-elbow_theta
	print(theta_centroid_real_life)
	
	targetx = camera_coords.x + (distance_camera_centroid*math.cos(theta_centroid_real_life))
	targety = camera_coords.y + (distance_camera_centroid*math.sin(theta_centroid_real_life))


	target_coords = Point(targetx,targety)
	return target_coords


# Calculate the distance between the base and the target coordinates
def calculate_distance_to_target(x, y):
	distance_from_base_to_target = math.sqrt(x**2 + y**2)  # Pythagoras' theorem
	return distance_from_base_to_target

# Calculate the needed rotation of the gripper to grab the object while remaining perpendicular to it
def calculate_gripper_angle(vx, vy):
	# Perpendicular direction vector
	px = -vy
	py = vx

	# Calculate the angle with respect to the x-axis
	angle_radians = math.atan2(py, px)
	angle_degrees = angle_radians * (180.0 / math.pi)

	# Normalize angle to be within the range of [-180, 180]
	if angle_degrees > 180:
		angle_degrees -= 360
	elif angle_degrees < -180:
		angle_degrees += 360

	# Account for the dead zone of the servos (<-150 and >150)
	if angle_degrees > 150:
		angle_degrees -= 180  # Flip 180 degrees
	elif angle_degrees < -150:
		angle_degrees += 180  # Flip 180 degrees

	# Normalize angle again to be within the range of [-150, 150]
	if angle_degrees > 150:
		angle_degrees -= 360
	elif angle_degrees < -150:
		angle_degrees += 360

	return angle_degrees

# Calculate the angles of the robot's elbow and base
def calculate_arm_angles(x, y, distance_to_target):  # x and y of the center of mass of the object, along with the distance
	arm_angles = np.zeros(2)
	# Use trigonometry to calculate the angles
	arm_angles[0] = 180 - (math.acos((length_ulna**2 + length_humerus**2 - distance_to_target**2) / (2 * length_humerus * length_ulna))) * (180 / math.pi)  # Elbow
	arm_angles[1] = (math.acos((distance_to_target**2 + length_humerus**2 - length_ulna**2) / (2 * length_humerus * distance_to_target)) + math.atan(abs(y) / abs(x))) * (180 / math.pi)  # Base angle
	return arm_angles

# Run all the previous calculations
def calculator(x, y, vx, vy):
	results = np.zeros(3)
	#print(f"external c++ calculator has run with parameters {x} {y} {vx} {vy}")  # Debug log

	pointC = calculate_real_life_coordinates(x, y)  # The point where gripper is going to be heading
	#pointC = calculate_real_life_coordinates(320, 240)  # The point where gripper is going to be heading
	print(f"relative coords {x} {y}")
	print(f"absolute coords {round(pointC.x,1)} {round(pointC.y,1)}\n")
	distance_to_target = calculate_distance_to_target(pointC.x, pointC.y)  # Calculate distance from base to target coordinates
	# if distance_to_target > length_humerus + length_ulna or distance_to_target < length_humerus - length_ulna:
	#	 print("\nout of reach\n")  # Cannot reach the point, return previous, assume object didn't move
	#	 return results  # Results of previous iteration
	arm_angles = calculate_arm_angles(pointC.x, pointC.y, distance_to_target)  # Calculate required angles for the elbow and base servos
	results[0] = calculate_gripper_angle(vx, vy)  # Calculate angle for the gripper to be perpendicular to the scissors
	results[1] = arm_angles[0]  # Store the arm angle variables
	results[2] = arm_angles[1]
	return results
